### ISSUE TITLE
BUG: sparse.linalg.eigs: sorting behavior reliability

### ISSUE THREAD

**Issue by tylerjereddy** [scipy/scipy#9460]
Related to #9453 and discussions with @rc and @lobpcg. Short summary is that I'm still ok with forcing the sort order in the test because I don't see any connection between the recent `lobpcg` adjustments and the observed sorting issue -- looks to be some vendored `ARPACK` stuff not related to recent merges -- a new test added for other behavior just happens to pick up on this instability in `eigs()`. But would be good to sort out, no pun intended, eventually.

I did a bit of `git bisection` style analysis on the problematic `vals` array sort order returned by `vals, vecs = eigs(A, k=2)` in `test_eigs_consistency`. I used 500 repeat trials to measure the consistency in the sorted results over git commit history, where a "consistency" is measured as either real or complex result that is sorted to be equivalent to reverse order `np.array([[5., 4.]])` -- I think the "imaginary" component of the number is always reported in the results, but didn't distinguish, my focus was the sort order.

- on current master ( a905607) in 500 trials `eigs` produces the "correct" sort order **94.8 %** of the time
- when I merged the largest option lobpcg fix ( 43193b90 ) 500 trials of eigs() still only sorts with **94.8%** fidelity
- if I go back to some commit from July 2018 ( c61e36fd )  -- still **94.8 %**
- if I go back to some commit from 2017 ( 541c0ee31 ) -- still **94.8 %**
- if I go back to before some ARPACK bundling changes in 2016 ( 0fbef6d6 ) -- **92.0 %**

The specific script used during my tests, which could still be cleaned up a bit more to focus on just 1 case, etc.:

```python
import numpy as np
import scipy
import scipy.sparse
from scipy.sparse.linalg.eigen.lobpcg import lobpcg
from scipy.sparse.linalg import eigs
from scipy.sparse import spdiags
from numpy.testing import assert_allclose

correct_counter = 0
for n, atol in [[20, 1e-3], [5, 1e-8]]:
    for i in range(500):
        vals = np.arange(1, n+1, dtype=np.float64)
        A = spdiags(vals, 0, n, n)
        vals, vecs = eigs(A, k=2)
        print("vals:", vals)
        
        if np.allclose(vals, np.array([[5., 4.]])):
           correct_counter += 1

print("Percent correct result sorting:", correct_counter / 5.)
```

**Comment by lobpcg** [comment]
Shouldn't the code above use
`
  if np.allclose(vals, np.array([[n, n-1]])):
` 

I also do not understand your percentage calculation.

**Comment by tylerjereddy** [comment]
500 trials with `[5, 1e-8]`, ignore the 500 trials with the other input, and dividing by 500 trials and multiplying by 100 to get % is same as / 5.

Could be cleaned up as I said, but should be ok for checking.

**Comment by tylerjereddy** [comment]
But yeah, when someone gets serious on the debugging they'd probably want something more formal for checking -- I just kind of took the test @rc wrote and gutted for bisecting.

**Comment by lobpcg** [comment]
The observed erratic sorting behavior of eigs is apparently a feature, not a bug. The docs do not promise any specific sorting. 

Searching for sort at scipy\sparse\linalg\eigen\arpack\arpack.py finds 

                rd = np.round(d, decimals=_ndigits[self.tp])
                if self.which in ['LR', 'SR']:
                    ind = np.argsort(rd.real)
                elif self.which in ['LI', 'SI']:
                    # for LI,SI ARPACK returns largest,smallest
                    # abs(imaginary) why?
                    ind = np.argsort(abs(rd.imag))
                else:
                    ind = np.argsort(abs(rd))
                if self.which in ['LR', 'LM', 'LI']:
                    d = d[ind[-k:]]
                    z = z[:, ind[-k:]]
                if self.which in ['SR', 'SM', 'SI']:
                    d = d[ind[:k]]
                    z = z[:, ind[:k]]

that sorting at the interface level is only performed under special rare randomly appearing circumstances, otherwise the sorting is performed by ARPACK itself. These two sorting scenarios are apparently inconsistent with each other at least in some cases, resulting in random sorting behavior. 

https://www.mathworks.com/help/matlab/ref/eigs.html actually documents specific sorting. If sorting is desired, that may the rule to follow.

**Comment by tylerjereddy** [comment]
@lobpcg Thanks for helping with this! Would you agree that for now, given a likely attempt to branch `1.2.x` in a few hours, it is probably ok to have the `np.sort()` forcing in the unit test for now, and we can check this in greater detail later if reliable sorting is needed for some reason in the future?

It sounds like we've probably not had any solid guarantee of sorting in the past, so seems to me that the unit test should indeed enforce that to avoid stochastic failures.

**Comment by lobpcg** [comment]
Yes, I now fully agree. Please keep the tests as you have done, with internal sorting.

**Comment by rc** [comment]
OK, thanks for exploring the issue - FYI: i am getting also 94.8% with scipy 1.1.0.dev0+2225a87. I must have been lucky not to trigger the problem when working on #9352. Also in the #9453 discussion I wrongly assumed that the complex eigenvalues were returned by lobpcg(), but they were returned by eigs().

So +1 to using np.sort in the test. If once the sorting order of eigs() is resolved, it can be eventually removed.

**Comment by ilayn** [comment]
I still think 20 is a very small size for a sparse algorithm test though. There might be other problems that can show up when the size grows.

**Comment by lobpcg** [comment]
Is this resolved and changes merged, and released, so it can now be closed?

**Comment by ilayn** [comment]
@lobpcg I didn't follow but did we figure out the sorting and added the bigger size matrices?

**Comment by lobpcg** [comment]
Upon a second thought, this script failing is likely also related to eigs missing eigenvalues, depending on an initialization. This is a normal behavior for ARPACK (as well as for LOBPCG without preconditioning) - if the initial vector(s) nearly orthogonal to one of the target eigenvectors, it is likely not to be computed.

I propose that the issue is closed, since there is no constructive way to fix it.

**Comment by ilayn** [comment]
After the rewrite I can comment on this a bit more informed. The returning eigenvalues are rather collected by ARPACK, that is to say, they show up as the algorithm starts to hit them depending on where it starts the process with the random vector. So it can come 4, 5 or 5, 4 depending on the initial vector direction. 

However, now that we took out the random seeding mechanism outside ARPACK we can freeze the result by using a definite rng and the results are going to be reproducible regardless of the discovery order unpredictability. Hence I'll add this to #22748 to be closed issues list.
